src/ports/postgres/modules/graph/pagerank.py_in

```diff
@@ -158,44 +313,198 @@ def pagerank(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
+            unconverged = 1
         plpy.execute("""
-                DROP TABLE IF EXISTS {cur};
-                ALTER TABLE {message} RENAME TO {cur}
-            """.format(**locals()))
+                    DROP TABLE IF EXISTS {cur};
+                    ALTER TABLE {message} RENAME TO {cur}
+                """.format(**locals()))
+        if unconverged == 0:
+            break
 
-    plpy.execute("ALTER TABLE {cur} RENAME TO {out_table}".format(**locals()))
+    # If there still are some unconverged groups/(entire table), update results.
+    if grouping_cols:
+        if unconverged > 0:
+            if threshold != 0:
+                # We completed max_iters, but there are still some unconverged groups
+                # Update the result and summary tables for unconverged groups.
+                update_result_tables(temp_summary_table, iteration_num,
+                    summary_table, out_table, cur, grouping_cols_list, cur_unconv)
+            else:
+                # No group has converged. List of all group values are in
+                # distinct_grp_table.
+                update_result_tables(temp_summary_table, iteration_num,
+                    summary_table, out_table, cur, grouping_cols_list, distinct_grp_table)
+    else:
+        plpy.execute("""ALTER TABLE {table_name} RENAME TO {out_table}
+            """.format(table_name=cur, **locals()))
+        plpy.execute("""
+                INSERT INTO {summary_table} VALUES
+                ({iteration_num}+1);
+            """.format(**locals()))
 
     ## Step 4: Cleanup
-    plpy.execute("""
-        DROP TABLE IF EXISTS {0},{1},{2},{3};
-        """.format(out_cnts, edge_temp_table, cur, message))
+    plpy.execute("""DROP TABLE IF EXISTS {0},{1},{2},{3},{4},{5};
+        """.format(out_cnts, edge_temp_table, cur, message, cur_unconv,
+                    message_unconv))
+    if grouping_cols:
+        plpy.execute("""DROP TABLE IF EXISTS {0},{1},{2};
+            """.format(vertices_per_group, temp_summary_table,
+                distinct_grp_table))
     plpy.execute("SET client_min_messages TO %s" % old_msg_level)
 
+def update_result_tables(temp_summary_table, i, summary_table, out_table,
+    res_table, grouping_cols_list, cur_unconv, message_unconv=None):
+    """
+        This function updates the summary and output tables only for thouse
+        groups that have converged. This is found out by looking at groups
+        that appear in cur_unvonv but not in message_unconv.
+        If this function is called after max_iter is completed, without
+        convergence, all the unconverged groups from cur_unconv is used
+        (note that message_unconv is renamed to cur_unconv before checking
+        for unconverged==0 in the pagerank function's for loop)
+    """
+    if message_unconv is None:
+        plpy.execute("""
+            DROP TABLE IF EXISTS {temp_summary_table};
+            CREATE TABLE {temp_summary_table} AS
```

Comment:
We can overwrite the temp_summary_table variable instead of creating a new table (as long as we make sure not to drop it at the end).